File: scripts/verify-annotatie-plaatsing.py
"""Controleert of de annotaties op dezelfde plek staan als in een onafhankelijke
referentie-render.

Achtergrond: bij pagina's met /Rotate is een terugkerende klacht dat de
annotaties gedraaid staan. Een screenshot alleen bewijst niets, want de app
maskeert een verkeerd ingelezen appearance-stream door hem bij het laden te
corrigeren. Deze controle vergelijkt daarom de annotatielaag van de app met de
annotatie-inkt die pypdfium2 zelf op de pagina zet.

Maat: welk deel van de referentie-annotatie-inkt de app op diezelfde plek
tekent. Correct is >= 80%. Ligt de dekking na 90 graden draaien duidelijk
hoger, dan staat de laag daadwerkelijk gedraaid.

Gebruik:
    python scripts/verify-annotatie-plaatsing.py [uitvoermap] [pdf-map]

De uitvoermap is die van verify-annotatie-plaatsing.mjs (standaard
annotatie-plaatsing). Per bestand komt er een verschilbeeld bij: rood is alleen
referentie, groen alleen app, blauw beide.
"""
import json
import os
import sys

import numpy as np
import pypdfium2 as pdfium
from PIL import Image, ImageChops

UIT = sys.argv[1] if len(sys.argv) > 1 else "annotatie-plaatsing"
MAP = sys.argv[2] if len(sys.argv) > 2 else r"C:\Users\rickd\Documents\GitHub\verification-files\PDF-bestanden"

DREMPEL_GOED = 0.80
DREMPEL_GEDRAAID = 0.15  # hoeveel beter de gedraaide variant moet passen
SCHAAL = 0.4


def referentiemasker(pad, slug):
    """Annotatie-inkt isoleren door met en zonder annotaties te renderen."""
    pagina = pdfium.PdfDocument(pad)[0]
    met = pagina.render(scale=SCHAAL, draw_annots=True).to_pil().convert("RGB")
    zonder = pagina.render(scale=SCHAAL, draw_annots=False).to_pil().convert("RGB")
    masker = ImageChops.difference(met, zonder).convert("L").point(lambda v: 255 if v > 24 else 0)
    masker.save(os.path.join(UIT, f"{slug}__ref.png"))
    return masker


def appmasker(rij, doelformaat):
    """De annotatielaag van de app uitsnijden op het paginavak en opschalen."""
    im = Image.open(rij["png"]).convert("RGBA")
    zoom, ox, oy = rij["zoom"], rij["ox"], rij["oy"]
    vak = (
        max(0, int(ox)),
        max(0, int(oy)),
        min(im.width, int(ox + rij["pageW"] * zoom)),
        min(im.height, int(oy + rij["pageH"] * zoom)),
    )
    if vak[2] - vak[0] < 5 or vak[3] - vak[1] < 5:
        return None
    laag = im.crop(vak).split()[3].point(lambda v: 255 if v > 40 else 0)
    return laag.resize(doelformaat, Image.BILINEAR).point(lambda v: 255 if v > 60 else 0)


def dekking(masker_a, masker_b):
    return float((masker_a & masker_b).sum()) / masker_a.sum() if masker_a.sum() else None


rapport = json.load(open(os.path.join(UIT, "rapport.json"), encoding="utf-8"))
regels = []

for rij in rapport:
    naam = rij["bestand"]
    if not rij.get("png") or not os.path.exists(rij["png"]):
        continue
    try:
        ref = referentiemasker(os.path.join(MAP, naam), rij["slug"])
        app = appmasker(rij, ref.size)
        if app is None:
            regels.append((naam, rij.get("appAnnots"), None, None, "paginavak buiten canvas"))
            continue
        a = np.array(ref, dtype=bool)
        b = np.array(app, dtype=bool)
        if not a.any():
            regels.append((naam, rij.get("appAnnots"), None, None, "geen referentie-inkt"))
            continue
        gedraaid = np.array(
            app.rotate(90, expand=True).resize(ref.size, Image.BILINEAR).point(lambda v: 255 if v > 60 else 0),
            dtype=bool,
        )
        regels.append((naam, rij.get("appAnnots"), dekking(a, b), dekking(a, gedraaid), None))

        verschil = np.zeros(a.shape + (3,), dtype=np.uint8)
        verschil[..., 0] = np.where(a & ~b, 255, 0)
        verschil[..., 1] = np.where(b & ~a, 255, 0)
        verschil[..., 2] = np.where(a & b, 200, 0)
        Image.fromarray(verschil).save(os.path.join(UIT, f"{rij['slug']}__vergelijk.png"))
    except Exception as fout:  # noqa: BLE001 - een onleesbaar bestand mag de sweep niet stoppen
        regels.append((naam, rij.get("appAnnots"), None, None, str(fout)[:70]))

print(f'{"bestand":50} {"annots":>6} {"dekking":>8} {"idem 90":>8}  oordeel')
print("-" * 96)
verdacht = []
for naam, aantal, d, d90, fout in regels:
    if fout:
        oordeel = f"niet meetbaar: {fout}"
    elif d >= DREMPEL_GOED:
        oordeel = "GOED"
    elif d90 and d90 > d + DREMPEL_GEDRAAID:
        oordeel = "GEDRAAID"
        verdacht.append(naam)
    else:
        oordeel = "AFWIJKEND"
        verdacht.append(naam)
    print(
        f"{naam[:50]:50} {str(aantal or '-'):>6} "
        f'{"-" if d is None else f"{d * 100:.1f}%":>8} '
        f'{"-" if d90 is None else f"{d90 * 100:.1f}%":>8}  {oordeel}'
    )

print()
print("aandacht nodig:", ", ".join(verdacht) if verdacht else "geen")
sys.exit(1 if verdacht else 0)

File: scripts/verify-annotatie-plaatsing.mjs
// Legt per verificatie-PDF de annotatielaag van de app vast, zodat
// verify-annotatie-plaatsing.py die kan vergelijken met een onafhankelijke
// referentie-render. Bedoeld voor de terugkerende melding "de annotaties staan
// gedraaid" op pagina's met /Rotate.
//
// Voorwaarden:
//   - Vite draait (npm run dev in open-pdf-studio/)
//   - De app draait met --mcp-server --mcp-port <poort>
//   - De WebView is bereikbaar via CDP (WEBVIEW2_ADDITIONAL_BROWSER_ARGUMENTS=
//     --remote-debugging-port=<poort>)
//
// Gebruik:
//   node scripts/verify-annotatie-plaatsing.mjs [pdf-map] [uitvoermap]
// Standaard: ../verification-files/PDF-bestanden en ./annotatie-plaatsing

import fs from 'fs';
import path from 'path';

const MCP = process.env.MCP_URL || 'http://127.0.0.1:9223/mcp';
const CDP_POORT = process.env.CDP_PORT || '9345';
const MAP = (process.argv[2] || 'C:/Users/rickd/Documents/GitHub/verification-files/PDF-bestanden').replace(/\\/g, '/');
const UIT = process.argv[3] || 'annotatie-plaatsing';

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const slug = (n) => n.replace(/\.pdf$/i, '').replace(/[^a-zA-Z0-9]+/g, '-').slice(0, 40);

let mcpId = 0;
async function mcp(naam, args = {}) {
  const res = await fetch(MCP, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json, text/event-stream' },
    body: JSON.stringify({ jsonrpc: '2.0', id: ++mcpId, method: 'tools/call', params: { name: naam, arguments: args } }),
  });
  const j = JSON.parse(await res.text());
  if (j.error) throw new Error(j.error.message || JSON.stringify(j.error));
  const c = j?.result?.content?.[0];
  if (c?.type === 'text') {
    try { return JSON.parse(c.text); } catch { return c.text; }
  }
  return j.result;
}

let ws = null;
let cdpId = 0;
async function cdpVerbind() {
  const lijst = await (await fetch(`http://127.0.0.1:${CDP_POORT}/json/list`)).json();
  const page = lijst.find((t) => t.type === 'page');
  if (!page) throw new Error('geen CDP-pagina gevonden');
  ws = new WebSocket(page.webSocketDebuggerUrl);
  await new Promise((r) => ws.addEventListener('open', r, { once: true }));
}
function cdpEval(expr) {
  return new Promise((res, rej) => {
    const id = ++cdpId;
    const h = (ev) => {
      const m = JSON.parse(ev.data);
      if (m.id !== id) return;
      ws.removeEventListener('message', h);
      if (m.error) return rej(new Error(JSON.stringify(m.error)));
      if (m.result.exceptionDetails) return rej(new Error(JSON.stringify(m.result.exceptionDetails).slice(0, 200)));
      res(m.result.result.value);
    };
    ws.addEventListener('message', h);
    ws.send(JSON.stringify({ id, method: 'Runtime.evaluate', params: { expression: expr, returnByValue: true } }));
  });
}

const TOESTAND = `(() => {
  const vp = window.__pdfViewport;
  const c = document.getElementById('annotation-canvas');
  return JSON.stringify({ zoom: vp.zoom, ox: vp.offsetX, oy: vp.offsetY, pageW: vp.pageW, pageH: vp.pageH,
                          cw: c && c.width, ch: c && c.height });
})()`;

const INKT = `(() => {
  const c = document.getElementById('annotation-canvas');
  if (!c || !c.width) return 0;
  const d = c.getContext('2d').getImageData(0, 0, c.width, c.height).data;
  let n = 0;
  for (let i = 3; i < d.length; i += 4) if (d[i] > 40) n += 1;
  return n;
})()`;

fs.mkdirSync(UIT, { recursive: true });
await cdpVerbind();

const bestanden = fs.readdirSync(MAP).filter((f) => f.toLowerCase().endsWith('.pdf')).sort();
const rapport = [];

for (const naam of bestanden) {
  const rij = { bestand: naam, slug: slug(naam) };
  try {
    const r = await mcp('app_open_pdf', { path: `${MAP}/${naam}` });
    rij.openOk = !!r?.ok;
    if (!rij.openOk) {
      rij.fout = JSON.stringify(r).slice(0, 100);
    } else {
      await sleep(4000);
      // Documenten openen sinds 1.88 standaard in de doorlopende weergave,
      // maar deze meting leest het enkelpagina-annotatiecanvas uit. Expliciet
      // naar enkelpagina schakelen; zonder deze stap meet alles 0% dekking.
      await cdpEval(`(() => {
        const b = [...document.querySelectorAll('button,[role=button]')].find(x => (x.title || '') === 'Enkele pagina');
        if (b) b.click();
        return true;
      })()`).catch(() => {});
      await sleep(1500);
      await mcp('app_fit_page', {});
      await sleep(3500);
      const a = await mcp('app_list_annotations', {});
      rij.appAnnots = (a?.annotations || a || []).length;
      if (rij.appAnnots > 0) {
        // De annotatielaag wordt na het passend maken opnieuw getekend. Wachten
        // tot er inkt op staat, anders meet de vergelijking een leeg canvas.
        for (let poging = 0; poging < 10; poging += 1) {
          rij.inktPixels = await cdpEval(INKT);
          if (rij.inktPixels > 0) break;
          await sleep(1000);
        }
        Object.assign(rij, JSON.parse(await cdpEval(TOESTAND)));
        const b64 = await cdpEval(`document.getElementById('annotation-canvas').toDataURL('image/png').slice(22)`);
        rij.png = path.join(UIT, `${rij.slug}__annlaag.png`);
        fs.writeFileSync(rij.png, Buffer.from(b64, 'base64'));
      }
    }
  } catch (e) {
    rij.fout = String(e.message).slice(0, 120);
  }
  try { await mcp('app_close_tab', {}); } catch (_) { /* tab kan al weg zijn */ }
  await sleep(1200);
  console.log(`${rij.openOk ? 'ok  ' : 'FOUT'} annots=${String(rij.appAnnots ?? '-').padStart(4)}  ${naam}${rij.fout ? `  << ${rij.fout}` : ''}`);
  rapport.push(rij);
}

fs.writeFileSync(path.join(UIT, 'rapport.json'), JSON.stringify(rapport, null, 1));
console.log(`\nklaar — ${rapport.filter((r) => r.png).length} annotatielagen vastgelegd in ${UIT}`);
console.log(`vervolg: python scripts/verify-annotatie-plaatsing.py ${UIT}`);
process.exit(0);
